enhancements/shared-and-guaranteed-cpu-pools.md

```diff
@@ -0,0 +1,407 @@
+6. The kubelet advertises `openshift.io/shared-cpus` and `openshift.io/guaranteed-cpus` extended resources 
+on the node based on the shared/isolated CPUSets specified in the config file.
+7. Something schedules:
+   * a pod with the `target.workload.openshift.io/management` annotation in a namespace 
+with the `workload.openshift.io/allowed` management annotation. The admission hook ignores the pod as it will
+be handled by the managementcpusoverride admission hook.
+   * a pod with Burstable or BestEffort QoS. The admission hook modifies the pod,
+adding `openshift.io/shared-cpus` requests/limits for each container, matching the CPU requests.
+   * a pod with Guaranteed QoS. The admission hook modifies the pod,
+adding `openshift.io/guaranteed-cpus` requests/limits for any containers with whole CPU requests/limits and
+`openshift.io/shared-cpus` requests/limits for other containers, matching the CPU requests.
+8. The scheduler sees the new pod and finds available `openshift.io/shared-cpus` and/or 
+`openshift.io/guaranteed-cpus` resources on a node. The scheduler places the pod on the node.
+9. Repeat steps 7-8 until all pods are running.
+
+### API Extensions
+
+A new admission hook in the Kubernetes API Server within OpenShift will mutate pods when they are created
+to add `openshift.io/guaranteed-cpus` and/or `openshift.io/shared-cpus` requests/limits as described above.
+
+### Implementation Details/Notes/Constraints
+
+One constraint of this approach is that the feature scope is cluster wide. When the feature gate is
+enabled, the new admission hook will mutate all incoming pods, adding requests/limits for the new
+extended resources. This means that all nodes in the cluster must have a Performance Profile that
+includes shared CPUs - without that, no pods would be scheduled on that node. Note that the isolated
+CPUs partition would be optional and would only be required on nodes that were going to run
+pods that had containers with whole CPU requests/limits.
+
+### Risks and Mitigations
+
+Nothing yet...
+
+### Drawbacks
+
+The changes proposed to kubelet will require us to carry patches to the upstream version. The idea of
+splitting the CPUs used by kubelet into a pre-defined shared and guaranteed pool might be considered
+upstream, but the use of the extended resources to track these is unlikely to be accepted. The
+alternative of adding a new first class resource (e.g. cpu-guaranteed) instead of extended resources
+would have huge impacts to existing code and is unlikely to be accepted.
+
+Another drawback is that the number of shared and guaranteed CPUs can only be changed with a reboot. However,
+there is no alternative that avoids a reboot since much of the required kernel tuning is static and cannot
+be changed at runtime.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. How will this be enabled at the openshift apiserver level? It looks like you can create an admission hook 
```

Comment:
Are opeshift apiserver pods going to be mutated as well? 
What about pods that are running before the admission hook starts running (if there're any)?

Reply:
> I will update the enhancement to be clear that the Management Workload Partitioning feature must be enabled in order to use this new feature

Decoupling it with WP means that even if the third pool would be accepted in u/s sometimes in the future, that still forces this feature to be OCP specific.

But if you are going in that direction I wonder if instead of having mutation hook, you can extend the default scheduler  using the scheduler-plugin framework  [https://github.com/kubernetes-sigs/scheduler-plugins](https://www.google.com/url?q=https://github.com/kubernetes-sigs/scheduler-plugins&sa=D&source=docs&ust=1686669446332969&usg=AOvVaw3WQj5R8Zguic5soxi1yOpI) 

The new scheduler plugin sees the guaranteed and shared cpus extended resources published by Kubelet, calculates the pod request based on the pod's QoS and CPUs requests/limts and admit the pod if the node has enough resources.

The extension for the scheduler is a plugin that runs as a pod, annotated with Management Workload Partitioning and placed on the reserved CPUs as well.

Reply:
The advantage of having an extension to the scheduler is that it's very common to do such thing u/s and it's completely pluggable and  dynamic.
IOW, you don't need that to be installed along with the cluster as you're doing with the admission hook.

Reply:
Thanks Tal-or - can you explain a bit more how this would work as a scheduler plugin? Are you suggesting that the pods would no longer have requests/limits for the extended resources? If that is the case, then how does the scheduler know how many shared/guaranteed CPUs are consumed already on each node by the pods currently running? Looking at the scheduler plugin interface it seems the interfaces just get passed the pod being scheduled, but I really don't know anything about this. Which plugin interfaces would we use for this and how would they be able to make the right decisions about where to schedule the new pod?

Reply:
> Are you suggesting that the pods would no longer have requests/limits for the extended resources?
Yes.

> If that is the case, then how does the scheduler know how many shared/guaranteed CPUs are consumed already on each node by the pods currently running? 

For every new pod that needs to be schedule, the scheduler would list all the existing pods on the node, calculates which pod belongs to which pool (By checking the QoS and resources of each pod).

At this point, the scheduler has the information about how much resources were allocated from each pool, and it also knows how much `Allocatable` resources at each pool the node has (because Kubelet populates this information to the node via the extended resources).

In the last step the scheduler figures out to which pool the new pod is belong (like it did for the existing pods), checks whether there are enough available resources, by subtract the allocated resources from `Allocatable` and mark the node as suitable or not.

> Looking at the scheduler plugin interface it seems the interfaces just get passed the pod being scheduled

You can initialize a lister to list all the non-terminal pods on the nodes.

> Which plugin interfaces would we use for this

https://pkg.go.dev/k8s.io/kubernetes/pkg/scheduler/framework#FilterPlugin

Reply:
Thanks!!! I will look into this option as it seems promising. One issue though is if we don't add the requests/limits for the new extended resources to the pods, the "[oc describe node](https://github.com/kubernetes/kubectl/blob/dfcbd3349ea39a580084e6c5c70897df13551071/pkg/util/resource/resource.go#L34)" won't show the "Allocated resources" properly. I guess we could patch the kubectl code to calculate the values for the shared/guaranteed extended resources based on the QoS of each pod, but that feels extremely ugly. I think we do need to give the user a way to see how much shared/guaranteed CPU resources are in use on each node, but maybe we don't want to use extended resources if we go with the scheduler plugin option? Any thoughts on a different/better way to show this?

Reply:
> the "[oc describe node](https://github.com/kubernetes/kubectl/blob/dfcbd3349ea39a580084e6c5c70897df13551071/pkg/util/resource/resource.go#L34)" won't show the "Allocated resources" properly.

hmm, I think you'll still be able to see it with `oc get -o yaml node` because it's part of the `NodeStatus`

> but maybe we don't want to use extended resources if we go with the scheduler plugin option

I don't see a way to completely avoid that, because the scheduler would need that information explicitly from the node. it needs to know the initial division of the different pools. 

Unless, you're passing that information to the scheduler when it starts. you can provide the pool size as an argument  to the scheduler, and the pool size consistent unless you change the value in the performanceProfile which requires reboot to the node anyway.

Reply:
> > the "[oc describe node](https://github.com/kubernetes/kubectl/blob/dfcbd3349ea39a580084e6c5c70897df13551071/pkg/util/resource/resource.go#L34)" won't show the "Allocated resources" properly.
> 
> hmm, I think you'll still be able to see it with `oc get -o yaml node` because it's part of the `NodeStatus`
> 

It won't be in the NodeStatus - it only contains the allocatable and the capacity. [The code](https://github.com/kubernetes/kubectl/blob/dfcbd3349ea39a580084e6c5c70897df13551071/pkg/util/resource/resource.go#L34) that sums up the  pod requests/limits to display "Allocated resources" depends on the extended resources requests/limits in each pod, which will not be there with this approach.

> > but maybe we don't want to use extended resources if we go with the scheduler plugin option
> 
> I don't see a way to completely avoid that, because the scheduler would need that information explicitly from the node. it needs to know the initial division of the different pools.
> 
> Unless, you're passing that information to the scheduler when it starts. you can provide the pool size as an argument to the scheduler, and the pool size consistent unless you change the value in the performanceProfile which requires reboot to the node anyway. That could work.

This won't work, because we can have multiple performanceProfiles (for different machine config pools).

Reply:
>  to display "Allocated resources" 

Oh sorry my bad I read it wrong.
Then yes you're right this is a bit more tricky. 
The only thing I can think of is using device-plugin to represent the shared/guaranteed as devices, but this is cumbersome.
Another option is to use new/existing object to contain this information.



> This won't work, because we can have multiple performanceProfiles (for different machine config pools).

No biggie. You can provide a ConfigMap that holds shared/guaranteed pool information associate with each node selector (the node selector map PerformanceProfile to group of  nodes).

The scheduler will examine each node based on the relevant pool (node -> node-selector -> shared/guaranteed pool)

Reply:
Hmm... I like the scheduler plugin approach because it feels much more self-contained. But the problem is going to be observability. Since we wouldn't be creating the new extended resources for shared/guaranteed CPUs, we have two issues:
1. How to publish the number of shared/guaranteed cpus configured on each node.
2. How to show the user how many shared/guaranteed cpus are in use on each node.

For 1, a simple option would be to have kubelet add annotations to each node (instead of publishing extended resources). The scheduler plugin could then use those annotations to know what was available on each node. Something like:
openshift.io/shared-cpus: x
openshift.io/guaranteed-cpus: y

For 2, we need a way for the user to see how many shared/guaranteed CPUs are configured and in use on each node. If extended resources are used (as in our current proposal), this is all automatic (handled already by "oc describe node"). If not using extended resources, I can think of a couple options:
1. We could add a new "oc" command to display the shared/guaranteed CPUs configured (by looking at the annotations on each node) and the shared/guaranteed CPUs in use (by looking at the pods on each node and their QoS class). 
2. We could patch the existing "oc describe node" command to add in this information (calculated in the same way). But this feels ugly and it doesn't look like we actually patch the kubectl code today (from what I can tell - not 100% sure though).

It doesn't feel like either of these options for showing the shared/guaranteed CPUs is going to be acceptable, so maybe going back to the approach of using extended resources and mutating the pods in the admission hook is the best we can do?

Reply:
> 2. How to show the user how many shared/guaranteed cpus are in use on each node.
We may kill two birds with one stone, if we'll manage to address https://github.com/openshift/enhancements/pull/1421#discussion_r1230103699 
Because RTE already presents in-used resources.

Reply:
I added a "New Scheduler Plugin" alternative to sum up this discussion.